### PR TITLE
WIP: Page Specific JS Route Creation

### PR DESCRIPTION
**Pull request by kave**
WIP

@anselmbradford
@rosskarchner  

Example Json file created on page save/edit
```
{
  "page_3": {
    "elements": [
      "well",
      "text_introduction"
    ],
    "slug": "demo-purposes"
  },
  "page_4": {
    "elements": [
      "text_introduction",
      "well"
    ],
    "slug": "landing-page"
  }
}
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -105,3 +105,8 @@ coverage.xml
 # IDE Specifc
 ############
 *.idea
+
+
+# Application Specifc
+############
+pages.json
```
Comment (anselmbradford) [inline .gitignore]: This is actually placed at `cfgov/v1/pages.json`. I'll update it.

File: cfgov/v1/wagtail_hooks.py [hunk 2]
```diff
@@ -33,3 +36,42 @@ def check_request_site(page, request, serve_args, serve_kwargs):
         if isinstance(page, CFGOVPage):
             if not page.shared:
                 raise Http404
+
+
+PAGES_JSON = settings.PROJECT_ROOT.child('v1', '_generated') + '/pages.json'
+
+def update_key_pages_json(request, page):
+    page = page.specific
+    data = json.load(open(PAGES_JSON))
```
Comment (anselmbradford) [inline cfgov/v1/wagtail_hooks.py]: Why is the page loading the existing state from the JSON? Why not write this file anew each time save is hit? Then we can gitignore this file.

**Comment by anselmbradford** [comment]
@kave https://github.com/cfpb/cfgov-refresh/commit/8bca7aa4580d0eafd4107569c2b6621ccfb5359d Adds a `buildScriptsRoutes` task to gulp that is run when `gulp scripts` is run, this task:
- Reads wagtail generated JSON.
- Parses pages in JSON and looks for elements that are in a lookup table as having associated JS.
- Write a JS page in the routes directory that searches the page dom for instances of a particular element and instantiates and initializes a new element of that type.
- Webpack scoops up generated routes like it usually does and works out require statements and bundling.

### To Test:
- Create a page in Wagtail, add a Well or whatever, and click save to generate `pages.json`.
- Manually edit `pages.json` and add `expandable` or `expandable_group` to the list of elements on the page.
- Run `gulp scripts`
- Click preview on the page in the Wagtail and look at the console output. It should say `Instantiated new Expandable` (or Group).

### Todo:
- Instead of JSON, ideally Wagtail should spit out the JS files in the routes that gulp is currently creating.
  
  `Wagtail > write JSON > gulp buildScriptsRoutes > read JSON > write JavaScript file > gulp scripts`
  
  vs
  
  `Wagtail > write JavaScript file > gulp scripts`.
- Currently there's a lookup table that shows whether an element is an molecule or organism and whether it has associated JavaScript. Instead this information should be inferred from the filesystem makeup. Whether an element is in the molecule or organism directory should determine its type and whether their is an associated JS file should determine if it needs inclusion in the page script.
- Page scripts will be created for the pages that don't have elements that have scripts themselves. This page will have the webpack loader and `'use strict;', but functionally nothing else. Pages that don't require any atomic element JS should not generate a  page script.
- Not currently deleting paths that have been deleted on the Wagtail side. The `unprocessed/js/routes/` directory will need to be cleared, **except** for the hardcoded paths that are used by the static routes (those from the existing site that aren't dynamically created by Wagtail).

### Caveats:
- This expects elements with an associated JS script to be able to be instantiated through `new` and to have an `init` method that takes no arguments. If an element needs initialization parameters these'll need to be set through attributes on the HTML or some other logic in the build process will need to be added.

File: cfgov/v1/wagtail_hooks.py (hunk 2 above)
Comment (kave) [inline cfgov/v1/wagtail_hooks.py]: @anselmbradford oh so it maintains a history of the pages that have been created, so on page delete i could just refer to the ID and remove it from the dict.. 
but now that i think about it, i could just add a delete flag so gulp knows to remove that path.. e.g
```
{
  "page_3": {
    "deleted": "true"
    "slug": "demo-purposes"
  }
}
```

@anselmbradford though i wonder if creating the page every-time might cause concurrency issues

File: cfgov/jinja2/v1/_includes/atoms/hyperlink.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
+{# ==========================================================================
```
Comment (kave) [inline cfgov/jinja2/v1/_includes/atoms/hyperlink.html]: ignore from https://github.com/cfpb/cfgov-refresh/pull/1198

File: cfgov/jinja2/v1/_includes/molecules/expandable.html [hunk 4]
```diff
@@ -11,29 +11,29 @@
    label:                Label you want on the Expandable.
```
Comment (kave) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: ignore from https://github.com/cfpb/cfgov-refresh/pull/1198

File: cfgov/jinja2/v1/_includes/organisms/expandable-group.html [hunk 5]
```diff
@@ -8,39 +8,49 @@
```
Comment (kave) [inline cfgov/jinja2/v1/_includes/organisms/expandable-group.html]: ignore from https://github.com/cfpb/cfgov-refresh/pull/1198

File: cfgov/jinja2/v1/_includes/atoms/hyperlink.html (hunk 3 above)
Comment (kave) [inline cfgov/jinja2/v1/_includes/atoms/hyperlink.html]: related to Expandables/Expandables Group atomic Element

File: cfgov/jinja2/v1/_includes/molecules/expandable.html (hunk 4 above)
Comment (kave) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: related to Expandables/Expandables Group atomic Element

File: cfgov/jinja2/v1/_includes/organisms/expandable-group.html (hunk 5 above)
Comment (kave) [inline cfgov/jinja2/v1/_includes/organisms/expandable-group.html]: related to Expandables/Expandables Group atomic Element

File: cfgov/jinja2/v1/browse-basic/index.html [hunk 6]
```diff
@@ -16,81 +16,102 @@
     <div class="block
```
Comment (kave) [inline cfgov/jinja2/v1/browse-basic/index.html]: related to Expandables/Expandables Group atomic Element

File: cfgov/jinja2/v1/wagtail/demo/index.html [hunk 7]
```diff
@@ -1,4 +1,4 @@
-{% extends "base.html" %}
+{% extends 'layout-side-nav.html' %}
```
Comment (kave) [inline cfgov/jinja2/v1/wagtail/demo/index.html]: related to Expandables/Expandables Group atomic Element

File: cfgov/v1/migrations/0012_auto_20151207_2022.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
```
Comment (kave) [inline cfgov/v1/migrations/0012_auto_20151207_2022.py]: related to Expandables/Expandables Group atomic Element

File: cfgov/v1/models/atoms.py [hunk 9]
```diff
@@ -33,6 +33,10 @@ def clean(self, data):
         else:
```
Comment (kave) [inline cfgov/v1/models/atoms.py]: related to Expandables/Expandables Group atomic Element

File: cfgov/v1/models/demo.py [hunk 10]
```diff
@@ -23,13 +23,15 @@ class DemoPage(CFGOVPage):
         ('image_text_5050', molecules.ImageText5050()),
```
Comment (kave) [inline cfgov/v1/models/demo.py]: related to Expandables/Expandables Group atomic Element

File: cfgov/v1/models/molecules.py [hunk 11]
```diff
@@ -8,7 +8,7 @@
```
Comment (kave) [inline cfgov/v1/models/molecules.py]: related to Expandables/Expandables Group atomic Element

File: cfgov/v1/models/organisms.py [hunk 12]
```diff
@@ -68,3 +68,17 @@ class MainContactInfo(blocks.StructBlock):
     class Meta:
```
Comment (kave) [inline cfgov/v1/models/organisms.py]: related to Expandables/Expandables Group atomic Element

File: cfgov/v1/wagtail_hooks.py [hunk 13]
```diff
@@ -1,10 +1,13 @@
-import os
+import os, json, itertools
```
Comment (anselmbradford) [inline cfgov/v1/wagtail_hooks.py]: Is the `json` import still needed?

File: package.json [hunk 14]
```diff
@@ -14,6 +14,7 @@
     "chai": "^3.3.0",
     "dateformat": "^1.0.11",
     "del": "^2.0.0",
+    "event-stream": "^3.0.5",
```
Comment (anselmbradford) [inline package.json]: This can be removed since the JS is built on the server.

File: cfgov/v1/util/js_routes.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    content = "'use strict';"
+
+    dir_depth = data['slug'].count('/')
+    path_depth = ''
+    for x in range(0, dir_depth):
+        path_depth += '../'
+
+    for idx, element in enumerate(elements):
+        class_name = util.to_camel_case(element)
+        type = LOOKUP_TABLE[class_name]
```
Comment (anselmbradford) [inline cfgov/v1/util/js_routes.py]: This is erroring because `class_name` is getting the name `ExpandablesGroup` instead of `ExpandableGroup`. It should be singular to align with design wiki.

File: cfgov/v1/wagtail_hooks.py (hunk 13 above)
Comment (kave) [inline cfgov/v1/wagtail_hooks.py]: ah nope, i'll remove it

**Comment by rosskarchner** [comment]
It's going to take me some time to grok all of this-- and I'm almost certainly not yet understating the benefits of the javascript routes.

But my first thought is, would it solve the same problem if we reflect all of the fields on a page, and reflect that as CSS classes on the body?  So, a page that contains a "thingamabob", could automatically have something like:

```html
<body class="has-thingamabob">
```

Of course, at that point, we could probably just as easily add a <script> tag for each field type found, which is even more direct.

**Comment by kave** [comment]
@anselmbradford could better explain the next benefits of js routes

File: cfgov/jinja2/v1/_includes/atoms/hyperlink.html [hunk 16]
```diff
@@ -1 +1,30 @@
...
+
+   value.text:    A string for the text of the link.
+
+   ========================================================================== #}
+
+{% if not value %}
+    {% do global_dict.update(
+        {
+            'hyperlink': {
+                'url' :       '/some-other-url/',
```
Comment (jimmynotjim) [inline cfgov/jinja2/v1/_includes/atoms/hyperlink.html]: Left align the values here

File: cfgov/jinja2/v1/_includes/molecules/expandable.html [hunk 17]
```diff
@@ -11,29 +11,29 @@
...
                          Default is false.
-   settings.is_expanded: Whether the Expandable is expanded or not.
+   value.is_expanded: Whether the Expandable is expanded or not.
                          Default is false.
    ========================================================================== #}
 
-{% macro render(label="", settings) %}
-{# Add the __expanded modifier class by default so all expandables are open
-   when JS is switched off, but add a data-js-hook to tell JS to remove it. #}
+{% if not value %}
```
Comment (jimmynotjim) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: Would it be possible to break out the data prototyping from the JS route stuff and into a new PR? Doing both sorta breaks the "Changes are limited to a single goal (no scope creep)" rule

Comment (jimmynotjim) [inline cfgov/jinja2/v1/_includes/molecules/expandable.html]: NVM, now I'm seeing that 1198 was included in here. Lets get that merged so it's clear what the changes are here.

**Comment by kave** [comment]
closed, we decided to go with another approach https://github.com/cfpb/cfgov-refresh/pull/1253
